File: apps/mobile/test/session-start.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import { SessionStartClient } from "../src/session-start.ts";
import { InvalidRealtimeResponseError, RealtimeApiError } from "../src/transport.ts";

const result = {
  id: "s1",
  account_id: "account-1",
  status: "active",
  audio_config: {
    codec: "opus",
    sample_rate_hz: 48000,
    channels: 1,
    echo_cancellation: true,
    noise_suppression: true,
    auto_gain_control: true,
  },
  capabilities: {
    webrtc: true,
    data_channel: true,
    microphone: true,
    speaker: true,
    speaker_diarization: true,
  },
  started_at: "2026-08-12T00:00:01Z",
  ended_at: null,
  created_at: "2026-08-12T00:00:00Z",
};

test("starts a session with an explicit initial mode and idempotency key", async () => {
  let requestInput = "";
  let requestInit: RequestInit = {};
  const client = new SessionStartClient({
    baseUrl: "https://api.example.test/",
    accessToken: async () => "access-1",
    createId: () => "start-1",
    fetchImpl: async (input, init) => {
      requestInput = input;
      requestInit = init ?? {};
      return new Response(JSON.stringify(result), { status: 200 });
    },
  });

  assert.deepEqual(await client.start("s1", "assistant"), result);
  assert.equal(requestInput, "https://api.example.test/api/v1/voice-sessions/s1/start");
  assert.equal(new Headers(requestInit.headers).get("Authorization"), "Bearer access-1");
  assert.equal(new Headers(requestInit.headers).get("Idempotency-Key"), "start-1");
  assert.deepEqual(JSON.parse(String(requestInit.body)), { initial_mode: "assistant" });
});

test("defaults assistant-capable callers to assistant", async () => {
  let body = "";
  const client = new SessionStartClient({
    baseUrl: "https://api.example.test",
    accessToken: "access-1",
    fetchImpl: async (_input, init) => {
      body = String(init?.body);
      return new Response(JSON.stringify(result), { status: 200 });
    },
  });

  await client.start("s1", undefined, "start-default");
  assert.deepEqual(JSON.parse(body), { initial_mode: "assistant" });
});

test("surfaces API errors and rejects a mismatched response", async () => {
  const rejected = new SessionStartClient({
    baseUrl: "https://api.example.test",
    accessToken: "access-1",
    fetchImpl: async () =>
      new Response(JSON.stringify({ error: { code: "runtime_operation_conflict" } }), {
        status: 409,
      }),
  });
  await assert.rejects(rejected.start("s1"), (error: unknown) =>
    error instanceof RealtimeApiError && error.code === "runtime_operation_conflict",
  );

  const mismatched = new SessionStartClient({
    baseUrl: "https://api.example.test",
    accessToken: "access-1",
    fetchImpl: async () =>
      new Response(JSON.stringify({ ...result, id: "another-session" }), { status: 200 }),
  });
  await assert.rejects(mismatched.start("s1"), InvalidRealtimeResponseError);

  const incomplete = new SessionStartClient({
    baseUrl: "https://api.example.test",
    accessToken: "access-1",
    fetchImpl: async () => {
      const { audio_config: _audioConfig, ...body } = result;
      return new Response(JSON.stringify(body), { status: 200 });
    },
  });
  await assert.rejects(incomplete.start("s1"), InvalidRealtimeResponseError);
});

File: sdks/device/session_start_test.go
package device

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func TestSessionStartClientSendsInitialMode(t *testing.T) {
	createdAt := time.Date(2026, 8, 12, 0, 0, 0, 0, time.UTC)
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/voice-sessions/session-1/start" || r.Method != http.MethodPost {
			t.Fatalf("request = %s %s", r.Method, r.URL.Path)
		}
		if r.Header.Get("Authorization") != "Bearer access-1" || r.Header.Get("Idempotency-Key") != "start-1" {
			t.Fatalf("headers = %#v", r.Header)
		}
		var request struct {
			InitialMode Mode `json:"initial_mode"`
		}
		if err := json.NewDecoder(r.Body).Decode(&request); err != nil || request.InitialMode != ModeAssistant {
			t.Fatalf("body = %#v, %v", request, err)
		}
		_ = json.NewEncoder(w).Encode(VoiceSessionStartResult{
			ID: "session-1", AccountID: "account-1", Status: VoiceSessionActive, CreatedAt: createdAt,
		})
	}))
	defer server.Close()

	client := &SessionStartClient{
		BaseURL: server.URL,
		AccessToken: AccessTokenSourceFunc(func(context.Context) (string, error) {
			return "access-1", nil
		}),
	}
	result, err := client.Start(context.Background(), "session-1", ModeAssistant, "start-1")
	if err != nil || result.Status != VoiceSessionActive {
		t.Fatalf("Start() = %#v, %v", result, err)
	}
}

func TestSessionStartClientDefaultsToInterpretation(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var request struct {
			InitialMode Mode `json:"initial_mode"`
		}
		_ = json.NewDecoder(r.Body).Decode(&request)
		if request.InitialMode != ModeInterpretation {
			t.Fatalf("initial mode = %q", request.InitialMode)
		}
		_ = json.NewEncoder(w).Encode(VoiceSessionStartResult{
			ID: "session-1", AccountID: "account-1", Status: VoiceSessionActive, CreatedAt: time.Now().UTC(),
		})
	}))
	defer server.Close()
	client := &SessionStartClient{BaseURL: server.URL, AccessToken: AccessTokenSourceFunc(func(context.Context) (string, error) {
		return "access-1", nil
	})}
	if _, err := client.Start(context.Background(), "session-1", "", "start-legacy"); err != nil {
		t.Fatal(err)
	}
}

func TestSessionStartClientSupportsDeviceSessionPath(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/device/voice-sessions/session-1/start" {
			t.Fatalf("path = %q, want device session path", r.URL.Path)
		}
		_ = json.NewEncoder(w).Encode(VoiceSessionStartResult{ID: "session-1", AccountID: "account-1", Status: VoiceSessionActive, CreatedAt: time.Now().UTC()})
	}))
	defer server.Close()
	client := &SessionStartClient{BaseURL: server.URL, SessionPath: "api/v1/device/voice-sessions", AccessToken: AccessTokenSourceFunc(func(context.Context) (string, error) { return "device-token", nil })}
	if _, err := client.Start(t.Context(), "session-1", ModeInterpretation, "device-start-1"); err != nil {
		t.Fatal(err)
	}
}

func TestSessionStartClientRejectsErrorsAndMismatchedResponse(t *testing.T) {
	for _, test := range []struct {
		name string
		body string
		code int
		want error
	}{
		{name: "API conflict", body: `{"error":{"code":"runtime_operation_conflict"}}`, code: http.StatusConflict, want: ErrRuntimeOperationConflict},
		{name: "mismatched session", body: `{"id":"other","account_id":"account-1","status":"active","started_at":null,"ended_at":null,"created_at":"2026-08-12T00:00:00Z"}`, code: http.StatusOK, want: ErrInvalidResponse},
	} {
		t.Run(test.name, func(t *testing.T) {
			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
				w.WriteHeader(test.code)
				_, _ = w.Write([]byte(test.body))
			}))
			defer server.Close()
			client := &SessionStartClient{BaseURL: server.URL, AccessToken: AccessTokenSourceFunc(func(context.Context) (string, error) {
				return "access-1", nil
			})}
			if _, err := client.Start(context.Background(), "session-1", ModeInterpretation, "start-1"); !errors.Is(err, test.want) {
				t.Fatalf("error = %v, want %v", err, test.want)
			}
		})
	}
}
